File: clients/typescript/test/jwt.test.ts
import { describe, it, expect } from "vitest";
import { decodeJwtPayload, isTokenExpired } from "../src/jwt.js";

function makeJwt(payload: Record<string, unknown>): string {
  const b64url = (o: unknown) =>
    Buffer.from(JSON.stringify(o)).toString("base64url");
  return `${b64url({ alg: "HS256", typ: "JWT" })}.${b64url(payload)}.sig`;
}

describe("jwt", () => {
  it("decodes the payload segment", () => {
    const token = makeJwt({ id: "u1", collection: "users", exp: 9999999999 });
    const p = decodeJwtPayload(token);
    expect(p?.id).toBe("u1");
    expect(p?.collection).toBe("users");
  });

  it("returns null for malformed tokens", () => {
    expect(decodeJwtPayload("not-a-jwt")).toBeNull();
    expect(decodeJwtPayload("")).toBeNull();
  });

  it("detects expiry with leeway", () => {
    const now = Math.floor(Date.now() / 1000);
    expect(isTokenExpired(makeJwt({ exp: now - 10 }))).toBe(true);
    expect(isTokenExpired(makeJwt({ exp: now + 3600 }))).toBe(false);
    expect(isTokenExpired(makeJwt({ exp: now + 5 }), 30)).toBe(true); // leeway pushes it past
    expect(isTokenExpired("garbage")).toBe(true);
  });
});

File: clients/dart/test/jwt_test.dart
import 'dart:convert';

import 'package:test/test.dart';
import 'package:zigbase_client/zigbase_client.dart';

String _makeJwt(Map<String, dynamic> payload) {
  final header =
      base64Url.encode(utf8.encode(jsonEncode({'alg': 'HS256', 'typ': 'JWT'})));
  final body = base64Url.encode(utf8.encode(jsonEncode(payload)));
  return '$header.$body.sig';
}

void main() {
  group('decodeJwtPayload', () {
    test('decodes the payload segment of a hand-built token', () {
      final segment =
          base64Url.encode(utf8.encode('{"id":"u1","exp":9999999999}'));
      final token = 'header.$segment.sig';
      final p = decodeJwtPayload(token);
      expect(p, isNotNull);
      expect(p?['id'], 'u1');
      expect(p?['exp'], 9999999999);
    });

    test('decodes the payload segment', () {
      final token =
          _makeJwt({'id': 'u1', 'collection': 'users', 'exp': 9999999999});
      final p = decodeJwtPayload(token);
      expect(p?['id'], 'u1');
      expect(p?['collection'], 'users');
    });

    test('returns null for malformed tokens', () {
      expect(decodeJwtPayload('not-a-jwt'), isNull);
      expect(decodeJwtPayload(''), isNull);
      expect(decodeJwtPayload('a.b'), isNull);
      expect(decodeJwtPayload('a..c'), isNull);
      expect(decodeJwtPayload('a.!!!notbase64.c'), isNull);
    });
  });

  group('isTokenExpired', () {
    test('false for far-future exp', () {
      final now = DateTime.now().millisecondsSinceEpoch ~/ 1000;
      expect(isTokenExpired(_makeJwt({'exp': now + 3600})), isFalse);
    });

    test('true for past exp', () {
      final now = DateTime.now().millisecondsSinceEpoch ~/ 1000;
      expect(isTokenExpired(_makeJwt({'exp': now - 10})), isTrue);
    });

    test('true for missing exp', () {
      expect(isTokenExpired(_makeJwt({'id': 'u1'})), isTrue);
    });

    test('true for malformed token', () {
      expect(isTokenExpired('garbage'), isTrue);
    });

    test('leeway pushes a near-future exp into expired', () {
      final now = DateTime.now().millisecondsSinceEpoch ~/ 1000;
      expect(isTokenExpired(_makeJwt({'exp': now + 5}), leewaySeconds: 30),
          isTrue);
      expect(isTokenExpired(_makeJwt({'exp': now + 5})), isFalse);
    });
  });
}
